python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -0,0 +1,417 @@
+
+        # Retrieve computed moments sigma and centroid.
+        determinantRadius = shape.moments_sigma
+        centroidResult = shape.moments_centroid
+
+        # Subtract center if required by configuration.
+        if self.config.subtractCenter:
+            centroidResult.x -= centroid.getX()
+            centroidResult.y -= centroid.getY()
+
+        # Convert GalSim's `galsim.PositionD` to `lsst.geom.Point2D`.
+        centroidResult = Point2D(centroidResult.x, centroidResult.y)
+
+        # Populate the record with the centroid results.
+        record.set(self.centroidResultKey, centroidResult)
+
+        # Convert GalSim measurements to lsst measurements.
+        try:
+            # Create an ellipse for the shape.
+            ellipse = afwGeom.ellipses.SeparableDistortionDeterminantRadius(
+                e1=shape.observed_shape.e1,
+                e2=shape.observed_shape.e2,
+                radius=determinantRadius,
+                normalize=True,  # Fail if |e|>1.
+            )
+            # Get the quadrupole moments from the ellipse.
+            quad = afwGeom.ellipses.Quadrupole(ellipse)
+        except InvalidParameterError as error:
+            raise measBase.MeasurementError(error, self.INVALID_PARAM.number)
+
+        # Store the quadrupole moments in the record.
+        record.set(self.shapeKey, quad)
+
+        # Store the flux if required by configuration.
+        if self.config.addFlux:
+            record.set(self.fluxKey, shape.moments_amp)
+
+        # TODO: calculate errors in shape, centroid?
+
+    def fail(self, record, error=None):
+        # Docstring inherited.
+        self.flagHandler.handleFailure(record)
+        if error:
+            centroid = record.getCentroid()
+            self.log.debug(
+                "Failed to measure shape for %d at (%f, %f): %s",
+                record.getId(),
+                centroid.getX(),
+                centroid.getY(),
+                error,
+            )
+
+
+class HsmSourceMomentsConfig(HsmMomentsConfig):
+    """Configuration for HSM adaptive moments measurement for sources."""
+    badMaskPlanes = pexConfig.ListField[str](
+        doc="Mask planes used to reject bad pixels.", default=["BAD", "SAT"]
+    )
+
+
+@measBase.register("ext_shapeHSM_HsmSourceMoments")
+class HsmSourceMomentsPlugin(HsmMomentsPlugin):
+    """Plugin for HSM adaptive moments measurement for sources."""
+    ConfigClass = HsmSourceMomentsConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None):
+        super().__init__(config, name, schema, logName)
+        self.centroidResultKey = afwTable.Point2DKey.addFields(schema, name, "HSM source centroid", "pixel")
+        self.shapeKey = afwTable.QuadrupoleKey.addFields(
+            schema, name, "HSM source moments", afwTable.CoordinateType.PIXEL
+        )
+        if config.addFlux:
+            self.fluxKey = measBase.FluxResultKey.addFields(schema, name + "_Flux", "HSM source flux", "dn")
```

Comment:
The units for this quantity is not "dn" for sure. Leave it blank.

Reply:
Ah, correct! Can we use "count"?
https://github.com/lsst/meas_extensions_trailedSources/blob/3ddfee87b48b9533850c8cd877b2d95ab53c4cf1/python/lsst/meas/extensions/trailedSources/VeresPlugin.py#L110

Reply:
Removed it.